### DISCUSSION TITLE
Consider defaulting trait implementations to `closed` to avoid inconsistent behaviour when they're dispatched concretely vs. through a bound

### DISCUSSION THREAD

**Discussion by utaal** [verus-lang/verus#399] [Language design]
Reported by @jonhnet. He's worried that for a spec function in a trait:

```rust
trait Thing {
  spec fn foo(a: bool) -> bool;
}
```

the behaviour of an `open` implementation:

```rust
impl Thing for Something {
  open spec fn foo(a: bool) -> bool {
    !a
  }
}
```

would change depending on whether one's accessing `foo` via a concrete `Something` or via a `T: Thing`: in one case you'd have access to the definition, and in the other you wouldn't, which changes the verifier's behaviour, possibly unexpectedly.

**Comment by tjhance** [comment]
I'm not sure I entirely understand the concern here, but I posted this possibly related issue recently: https://github.com/verus-lang/verus/issues/382

**Comment by utaal** [comment]
[Verus all-hands]

This is not fixable by disallowing `open` / `close` qualifiers; it's intrinsic in how the traits are treated.
